test/Oscoin/Test/Data/OscoinTx.hs

```diff
@@ -0,0 +1,183 @@
+module Oscoin.Test.Data.OscoinTx where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Address (Address)
+import qualified Oscoin.Crypto.Hash as Crypto
+import           Oscoin.Crypto.PubKey (PublicKey)
+import           Oscoin.Data.Ledger
+import           Oscoin.Data.OscoinTx
+import           Oscoin.Test.Crypto
+import           Oscoin.Test.Util
+import           Test.Oscoin.P2P.Handshake (genKeyPair, genKeyPairPair)
+
+import qualified Crypto.Data.Auth.Tree as WorldState
+
+import           Test.Oscoin.Crypto.Address (genAddress)
+
+import           Test.Hedgehog.Extended
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+
+import           Hedgehog hiding (evalEither, forAll, (===))
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+
+tests :: forall c. Dict (IsCrypto c) -> [TestTree]
+tests Dict =
+    [ testGroup "Data.OscoinTx"
+        [ testProperty "Transfer balance"        (propTransferBalance @c Dict)
+        , testProperty "Apply transaction"       (propApplyTx         @c Dict)
+        ]
+    ]
+
+propApplyTx :: forall c. Dict (IsCrypto c) -> Property
+propApplyTx Dict = property $ do
+    (pk, pk') <- publicKeyPair @c
+
+    pk /== pk'
+
+    ws    <- forAll genWorldState
+    tx    <- mkTx <$> forAll (genRegisterProject pk)
```

Comment:
This can be an arbitrary transaction right? So why not use `genTx`?

Reply:
Not really, or I have to complicate the tests a lot: the reason I picked "register project", is that it cannot fail at the payload level - this way I'm only testing the failure scenarios at the Tx level.

Reply:
Maybe I should add a comment about this?

Reply:
Added comment. f561b6a